Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp

```diff
@@ -57,9 +57,14 @@ void MainThreadPermissionObserver::stateChanged(PermissionState newPermissionSta
 {
     m_state = newPermissionState;
 
-    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [permissionObserver = m_permissionObserver, newPermissionState](auto&) {
-        if (permissionObserver)
-            permissionObserver->stateChanged(newPermissionState);
+    if (m_source == PermissionQuerySource::Window && m_permissionStatus) {
+        m_permissionStatus->stateChanged(newPermissionState);
+        return;
+    }
+
+    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [permissionStatus = m_permissionStatus, newPermissionState](auto&) {
```

Comment:
You can call ScriptExecutionContext::ensureOnContextThread() to avoid dispatching when we're already on the main thread.